libyul/ControlFlowSideEffectsCollector.cpp

```diff
@@ -51,9 +51,15 @@ void ControlFlowBuilder::operator()(If const& _if)
 {
 	visit(*_if.condition);
 	ControlFlowNode* node = m_currentNode;
-	(*this)(_if.body);
+
+	ControlFlowNode* ifEnd = newNode();
+	node->successors.emplace_back(ifEnd);
+
 	newConnectedNode();
-	node->successors.emplace_back(m_currentNode);
+    (*this)(_if.body);
+	
+    m_currentNode->successors.emplace_back(ifEnd);
+    m_currentNode = ifEnd;
```

Comment:
You should add a test case in [`test/libyul/controlFlowSideEffects`](https://github.com/ethereum/solidity/tree/develop/test/libyul/controlFlowSideEffects) to show that these changes actually effect the side-effects.